# CS412 - Machine Learning - 2020

# Homework 2

100 pts

# Goal

The goal of this homework is to get familiar feature handling and cross validation.


# Dataset

German Credit Risk dataset, prepared by Prof. Hoffman, classifies each person as having a good or bad credit risk. The dataset that we use consists of both numerical and categorical features.



# Task

Build a k-NN classifier with scikit-learn library to classify people as bad or good risks for the german credit dataset. 

# Software

Documentation for the necessary functions can be accessed from the link below.

[http://scikit-learn.org/stable/supervised_learning.html](http://scikit-learn.org/stable/supervised_learning.html)

# Submission

Follow the instructions at the end.


# 1) Initialize

First, make a copy of this notebook in your drive

In [ ]:
# Mount to your drive, in this way you can reach files that are in your drive
# Run this cell
# Go through the link that will be showed below
# Select your google drive account and copy authorization code and paste here in output and press enter
# You can also follow the steps from that link
# https://medium.com/ml-book/simplest-way-to-open-files-from-google-drive-in-google-colab-fae14810674 

from google.colab import drive
drive.mount('/content/drive')

# 2) Load Dataset

To start working for your homework, take a copy of the folder, given in the below link to your own google drive. You find the train and test data under this folder.

[https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing](https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing)

After copy the folder, copy the path of the train and test dataset to paste them in the below cell to load your data.


In [ ]:
import pandas as pd

train_df = pd.read_csv('/content/drive/My Drive/german_credit_train.csv')

test_df = pd.read_csv('/content/drive/My Drive/german_credit_test.csv')

train_df.head()
#test_df.head()

AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
0           A14        12           A32  ...               A143    A152    1
1           A11         9           A32  ...               A143    A152    1
2           A11        18           A34  ...               A143    A153    1
3           A11        14           A32  ...               A143    A152    2
4           A14        15           A32  ...               A143    A152    2

[5 rows x 13 columns]

# 3) Optional - Analyze the Dataset 

You can use the functions of the pandas library to analyze your train dataset in detail - **this part is OPTIONAL - look around the data as you wish**.


*   Display the number of instances and features in the train ***(shape function can be used)**
*   Display 5 random examples from the train ***(sample function can be used)**
*   Display the information about each features ***(info method can be used)**



In [ ]:
# Print shape
print("Train data dimensionality: ", train_df.shape )

# Print shape
print("Test data dimensionality: ", test_df.shape )

# Print random 5 rows
print("Examples from train data: "  )
train_df.sample(n=5)




Train data dimensionality:  (800, 13)
Test data dimensionality:  (200, 13)
Examples from train data: 


AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
144           A11        18           A32  ...               A143    A152    2
79            A14        18           A33  ...               A143    A152    1
663           A12        12           A34  ...               A143    A152    1
263           A11        24           A32  ...               A141    A153    1
791           A12         9           A32  ...               A143    A152    1

[5 rows x 13 columns]

In [ ]:
# Print the information about the dataset
print("Information about train data ", train_df.info )

Information about train data  <bound method DataFrame.info of     AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
0             A14        12           A32  ...               A143    A152    1
1             A11         9           A32  ...               A143    A152    1
2             A11        18           A34  ...               A143    A153    1
3             A11        14           A32  ...               A143    A152    2
4             A14        15           A32  ...               A143    A152    2
..            ...       ...           ...  ...                ...     ...  ...
795           A12         9           A32  ...               A143     NaN    1
796           A14        12           A34  ...               A143     NaN    1
797           A12        36           A32  ...               A143    A153    2
798           A12        12           A32  ...               A143    A152    1
799           A13         9           A32  ...               A143    

# 4) Define your train and test labels

*  Define labels for both train and test data in new arrays 
*  And remove the label column from both train and test sets do tht it is not used as a feature! 


(**you can use pop method**)


In [ ]:
# Define labels
train_label = train_df.pop('Risk')
test_label = test_df.pop('Risk')

In [ ]:
train_df.head()
#risk label is popped

AccountStatus  Duration CreditHistory  ...  Age OtherInstallPlans Housing
0           A14        12           A32  ...   38              A143    A152
1           A11         9           A32  ...   25              A143    A152
2           A11        18           A34  ...   36              A143    A153
3           A11        14           A32  ...   45              A143    A152
4           A14        15           A32  ...   40              A143    A152

[5 rows x 12 columns]

# 5) Handle missing values if any 

*   Print the columns that have **NaN** values (**isnull** method can be used)
*   You can impute missing values with mode of that feature or remove samples or attributes
*   To impute the test set, you should use the mode values that you obtain from **train** set, as **you should not be looking at your test data to gain any information or advantage.**



In [ ]:
# Print columns with NaN values
print("Number of rows: ", len(train_df))
print("Number of rows without nan values:", len(train_df.dropna()))
print("We see that 80 rows have nan values train dataframe.\n")

train_df.isnull().any()


Number of rows:  800
Number of rows without nan values: 720
We see that 80 rows have nan values train dataframe.



AccountStatus        False
Duration             False
CreditHistory        False
CreditAmount         False
SavingsAccount       False
EmploymentSince      False
PercentOfIncome      False
PersonalStatus       False
Property             False
Age                  False
OtherInstallPlans    False
Housing               True
dtype: bool

In [ ]:
print("We see that only the housing column has nan values and there are 80 rows with missing housing value.\n")
train_df.isnull().sum()


We see that only the housing column has nan values and there are 80 rows with missing housing value.



AccountStatus         0
Duration              0
CreditHistory         0
CreditAmount          0
SavingsAccount        0
EmploymentSince       0
PercentOfIncome       0
PersonalStatus        0
Property              0
Age                   0
OtherInstallPlans     0
Housing              80
dtype: int64

In [ ]:
# Print columns with NaN values
print("Number of rows: ", len(test_df))
print("Number of rows without nan values:", len(test_df.dropna()))
print("We see that 20 rows have nan values in the test dataframe.\n")

test_df.isnull().any()


Number of rows:  200
Number of rows without nan values: 180
We see that 20 rows have nan values in the test dataframe.



AccountStatus        False
Duration             False
CreditHistory        False
CreditAmount         False
SavingsAccount       False
EmploymentSince      False
PercentOfIncome      False
PersonalStatus       False
Property             False
Age                  False
OtherInstallPlans    False
Housing               True
dtype: bool

In [ ]:
print("We see that only the housing column has nan values and there are 20 rows with missing housing value.\n")
test_df.isnull().sum()

We see that only the housing column has nan values and there are 20 rows with missing housing value.



AccountStatus         0
Duration              0
CreditHistory         0
CreditAmount          0
SavingsAccount        0
EmploymentSince       0
PercentOfIncome       0
PersonalStatus        0
Property              0
Age                   0
OtherInstallPlans     0
Housing              20
dtype: int64

In [ ]:
# Impute missing values by replacing with mode value
train_df['Housing'] = train_df['Housing'].fillna(train_df['Housing'].mode()[0])
test_df['Housing'] = test_df['Housing'].fillna(train_df['Housing'].mode()[0])
print("We imputed the housing values in train and test dataframes with the most common value in the train dataframe.")

We imputed the housing values in train and test dataframes with the most common value in the train dataframe.


# 6) Transform categorical / ordinal features

* Transform all categorical / ordinal features using the methods that you have learnt in lectures and recitation 4 for both train and test data
* You saw the dictionary use for mapping in recitation. (You can use **replace function** to assign new values to the categories of a column).

*  The class of the categorical attributes in the dataset are defined as follows:
  - Status of existing checking account
     - A11 :      ... <    0 DM
	- A12 : 0 <= ... <  200 DM
	- A13 :      ... >= 200 DM / salary assignments for at least 1 year
     - A14 : no checking account

 - Credit history
    - A30 : no credits taken/all credits paid back duly
    - A31 : all credits at this bank paid back duly
	- A32 : existing credits paid back duly till now
    - A33 : delay in paying off in the past
	- A34 : critical account/other credits existing (not at this bank)

  - Savings account
    - A61 :          ... <  100 DM
	- A62 :   100 <= ... <  500 DM
	- A63 :   500 <= ... < 1000 DM
	- A64 :          .. >= 1000 DM
    - A65 :   unknown/ no savings account

 - Employment Since
    - A71 : unemployed
    - A72 :       ... < 1 year
	- A73 : 1  <= ... < 4 years  
	- A74 : 4  <= ... < 7 years
	- A75 :       .. >= 7 years
 
 - Personal Status
    - A91 : male   : divorced/separated
	- A92 : female : divorced/separated/married
    - A93 : male   : single
	- A94 : male   : married/widowed
	- A95 : female : single

  - Property
     -  A121 : real estate
	- A122 : if not A121 : building society savings agreement/life insurance
    - A123 : if not A121/A122 : car or other, not in attribute 6
	- A124 : unknown / no property

 - OtherInstallPlans  
    - A141 : bank
	- A142 : stores
	- A143 : none

 - Housing
    -  A151 : rent
	 - A152 : own
	- A153 : for free

In [ ]:
# Transform the ordinal attributes
checking_account_map = {'A14':0, 'A11':1, 'A12':2, 'A13':3}
train_df['AccountStatus'] =train_df['AccountStatus'].replace(checking_account_map)
test_df['AccountStatus'] =test_df['AccountStatus'].replace(checking_account_map)

credit_history_map = {'A30':4, 'A31':3, 'A32':2, 'A33':1, 'A34':0}
train_df['CreditHistory'] =train_df['CreditHistory'].replace(credit_history_map)
test_df['CreditHistory'] =test_df['CreditHistory'].replace(credit_history_map)

savings_account_map = {'A65':0, 'A61':1, 'A62':2, 'A63':3, 'A64':4}
train_df['SavingsAccount'] =train_df['SavingsAccount'].replace(savings_account_map)
test_df['SavingsAccount'] =test_df['SavingsAccount'].replace(savings_account_map)

employment_since_map = {'A71':0, 'A72':1, 'A73':2, 'A74':3, 'A75':4}
train_df['EmploymentSince'] =train_df['EmploymentSince'].replace(employment_since_map)
test_df['EmploymentSince'] =test_df['EmploymentSince'].replace(employment_since_map)

property_map = {'A124':0, 'A123':1, 'A122':2, 'A121':3}
train_df['Property'] =train_df['Property'].replace(property_map)
test_df['Property'] =test_df['Property'].replace(property_map)


In [ ]:
# Transform the categorical attributes #train_df['OtherInstallPlans'], train_df['Housing']
from sklearn.preprocessing import OneHotEncoder
enc= OneHotEncoder(handle_unknown = 'ignore')
dummy = enc.fit_transform(train_df[["OtherInstallPlans", "PersonalStatus", "Housing"]]).toarray()
dummy_df = pd.DataFrame(dummy)

train_df = pd.merge(train_df, dummy_df, right_index=True, left_index=True)
train_df = train_df.drop(columns=["OtherInstallPlans", "PersonalStatus", "Housing"])

In [ ]:
dummy_test = enc.transform(test_df[["OtherInstallPlans", "PersonalStatus", "Housing"]]).toarray()
dummy_test_df = pd.DataFrame(dummy_test)

test_df = pd.merge(test_df, dummy_test_df, right_index=True, left_index=True)
test_df = test_df.drop(columns=["OtherInstallPlans", "PersonalStatus", "Housing"])

# 7) Build a k-NN classifier on training data and perform models selection using 5 fold cross validation

*  Initialize k-NN classifiers with **k= 5, 10, 15**
*  Calculate the cross validation scores using cross_al_score method, number of folds is 5. 
*  Note: Xval is performed on training data! Do not use test data in any way and do not separate a hold-out validation set, rather use cross-validation.

Documentation of the cross_val_score method:

[https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score)

*  Stores the average accuracies of these folds
*  Select the value of k using the cross validation results. 

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(train_df, train_label, test_size=0.2, random_state=42)
x_train, x_valid, y_train, y_valid = train_test_split(x_train,y_train,test_size=0.25,random_state=42)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from statistics import mean
from sklearn.metrics import accuracy_score

# k values
kVals = [5,10,15]

# Save the accuracies of each value of kVal in [accuracies] variable
accuracies = []
# Loop over values of k for the k-Nearest Neighbor classifier
# Initialize a k-NN classifier with k neighbors
for k in kVals:
  knn = KNeighborsClassifier(n_neighbors=k)
  knn.fit(x_train,y_train)
  y_pred_knn = knn.predict(x_valid)
  accuracies.append(accuracy_score(y_valid,y_pred_knn))

  # Calculate the 5 fold cross validation scores using cross_val_score
  # cv parameter: number of folds, in our case it must be 5
  scores = cross_val_score(knn, x_train, y_train, cv=5)
  print("Accuracy for knn=", k ,": %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

  # Stores the average accuracies of the scores in accuracies variable, you can use mean method
print(accuracies)

Accuracy for knn= 5 : 0.66 (+/- 0.07)
Accuracy for knn= 10 : 0.68 (+/- 0.04)
Accuracy for knn= 15 : 0.68 (+/- 0.06)
[0.7, 0.7375, 0.75625]


# 8) Retrain using all training data and test on test set

* Train a classifier with the chosen k value of the best classifier using **all training data**. 

Note:  k-NN training involves no explicit training, but this is what we would do after model selection with decision trees or any other ML approach (we had 5 diff. models -one for each fold - for each k in the previous step - dont know which one to submit. Even if we picked the best one, it does not use all training samples.

* Predict the labels of testing data 

* Report the accuracy 

In [ ]:
from sklearn.metrics import accuracy_score

# Train the best classifier using all training set
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(train_df, train_label)
accuracy_train = cross_val_score(knn, train_df, train_label, cv=5)

print("Accuracy over all train data:", accuracy_train.mean())

# Estimate the prediction of the test data

knn.fit(test_df, test_label)
estimate = knn.predict(test_df)

# Print accuracy of test data
accuracy_test = accuracy_score(estimate, test_label)
print("Accuracy score of test data:", accuracy_test)

Accuracy over all train data: 0.7074999999999999
Accuracy score of test data: 0.705


# 9) Bonus (5pts)

There is a limited bonus for any extra work that you may use and improve the above results. 

You may try a larger k values, scale input features, remove some features, .... Please **do not overdo**, maybe spend another 30-60min on this. The idea is not do an exhaustive search (which wont help your understanding of ML process), but just to give some extra points to those who may look at the problem a little more comprehensively. 

**If you obtain better results than the above, please indicate the best model you have found and the corresponding accuracy.**

E.g. using feature normalization ..... and removing .... features and using a value k=...., I have obtained ....% accuracy.


In [ ]:
knn = KNeighborsClassifier(n_neighbors=15)
knn.fit(train_df, train_label)
accuracy_train = cross_val_score(knn, train_df, train_label, cv=5)

print("Accuracy over all train data:", accuracy_train.mean())

# Estimate the prediction of the test data

knn.fit(test_df, test_label)
estimate = knn.predict(test_df)

# Print accuracy of test data
accuracy_test = accuracy_score(estimate, test_label)
print("Accuracy score of test data:", accuracy_test)

Accuracy over all train data: 0.7100000000000001
Accuracy score of test data: 0.675


In [ ]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(train_df, train_label)
accuracy_train = cross_val_score(knn, train_df, train_label, cv=5)

print("Accuracy over all train data:", accuracy_train.mean())

# Estimate the prediction of the test data

knn.fit(test_df, test_label)
estimate = knn.predict(test_df)

# Print accuracy of test data
accuracy_test = accuracy_score(estimate, test_label)
print("Accuracy score of test data:", accuracy_test)

Accuracy over all train data: 0.6775
Accuracy score of test data: 0.705


# 10) Notebook & Report
We are going to examine the German Credit Risk Data. This dataset classifies people as good or bad credit risks. 
We have splitted the data by %80 train, %20 test.

We have found that only "housing" feature has missing values, and we have imputed these missing values with the most common housing value in the training set.

Some features were given with codes that has meanings, like "A30 : no credits taken/all credits paid back duly", we have given them values if they were ordinal; and categorized the categorical features with dummy variables.

The average accuracy scores turned out to be for k=5 -> 0.7, for k=10 -> 0.7375, and forr k=15 -> 0.756. 

With 5-cross validation, accuracy for knn= 5 was 0.66 with +/- 0.07 standard deviation, accuracy for knn= 10 was 0.68 with +/- 0.04 std dev. and accuracy for knn= 15 was 0.68 with +/- 0.06 std dev.

I picked the knn=10 since it gave %68 accuracy score with 5-cross validation with 0.06 std dev, which was the best result between these 5-10-15 neighbored models.

We have obtained the best results with the k-nearest-neighbor classifier with k=10 giving classification accuracy of %70.5 on test data.

We have tried k=5 and k=15 also, even though k=5 gave less accuracy score on training data, it gave the same accuracy score with k=10 on test data.
And as expected, k=15 had given bigger accuracy score on training data but have given %67.5 on test data, which may mean there is kind of an overfit.